### PR TITLE
Makefile: fix for GCC 7+ and clang 4+

### PR DESCRIPTION
**Pull request by choleraehyq**
@maysamyabandeh @IslamAbdelRahman PTAL

Fix https://github.com/facebook/rocksdb/issues/2672

Signed-off-by: Cholerae Hu <huyingqian@pingcap.com>

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -255,6 +255,19 @@ default: all
 WARNING_FLAGS = -W -Wextra -Wall -Wsign-compare -Wshadow \
   -Wno-unused-parameter
 
+CCVERSION = $(shell cc -dumpversion)
```
Comment (grooverdan) [inline Makefile]: cc -> $CXX
or ${CXX:-cc} though I have doubts that cc is right.

**Comment by siying** [comment]
Is it possible to fix the code rather than suppress the warning?

**Comment by ajkr** [comment]
@siying this should enable compiler support for allocating over-aligned types.

We still don't use `new` properly in older gcc versions. But I don't think it's critical as we over-align to protect against false cache-line sharing. So it should be a perf issue, not a correctness issue.

File: Makefile [hunk 2]
```diff
@@ -255,6 +255,19 @@ default: all
 WARNING_FLAGS = -W -Wextra -Wall -Wsign-compare -Wshadow \
   -Wno-unused-parameter
 
+CCVERSION = $(shell $(CXX) -dumpversion)
+CCNAME = $(shell $(CXX) --version | awk 'NR==1' | cut -f1 -d " ")
+
+ifeq ($(CCNAME), clang)
+ifeq ($(CCVERSION), 4*)
+	WARNING_FLAGS += -faligned-new
```
Comment (ajkr) [inline Makefile]: Can you put it in `CXX_FLAGS` instead as it's not really a warning flag?

**Comment by ajkr** [comment]
Sorry our tools won't let me pull it unless the CLA is signed (https://code.facebook.com/cla).

**Comment by choleraehyq** [comment]
Signed! @ajkr

**Pull request closed without merging**